### PR TITLE
Cura 11650 thumnails missing

### PR DESCRIPTION
**Pull request by saumyaj3**
# Description
CURA-11650

added creating snapshot if not created till 10 attempts

### CODE REVIEW

File: cura/Snapshot.py [hunk 1]
```diff
@@ -182,11 +193,19 @@ def snapshot(width = 300, height = 300):
...
-            except (ValueError, AttributeError):
-                Logger.logException("w", "Failed to crop the snapshot!")
+            if Snapshot._prereadSnapshot(pixel_output):
+                try:
+                    min_x, max_x, min_y, max_y = Snapshot.getImageBoundaries(pixel_output)
+                except (ValueError, AttributeError) as e:
+                    Logger.logException("w", f"Failed to crop the snapshot! {e}")
+                    return None
+            elif number_of_attempts == 0:
+                Logger.warning( f"Failed to crop the snapshot even after 10 attempts!")
```
Comment (casperlamboo) [inline cura/Snapshot.py:203]:
```suggestion
                Logger.warning( f"Failed to crop the snapshot even after {Snapshot.ATTEMPTS_FOR_SNAPSHOT} attempts!")
```

File: cura/Snapshot.py [hunk 2]
```diff
@@ -211,3 +230,8 @@ def snapshot(width = 300, height = 300):
             transformMode = QtCore.Qt.TransformationMode.SmoothTransformation)
 
         return scaled_image
+
+    @staticmethod
+    def _prereadSnapshot(snap = QImage):
```
Comment (casperlamboo) [inline cura/Snapshot.py:235]: Why default the `snap` arg to a `QImage`? didn't you mean to type the function in this case?
```suggestion
    def _prereadSnapshot(snap: QImage) -> bool:
```

File: cura/Snapshot.py [hunk 3]
```diff
@@ -182,11 +193,19 @@ def snapshot(width = 300, height = 300):
...
+                except (ValueError, AttributeError) as e:
+                    Logger.logException("w", f"Failed to crop the snapshot! {e}")
+                    return None
+            elif number_of_attempts == 0:
+                Logger.warning( f"Failed to crop the snapshot even after 10 attempts!")
                 return None
+            else:
+                number_of_attempts = number_of_attempts - 1
+                Logger.info("Trying to get the snapshot again.")
+                return Snapshot.snapshot(width, height, number_of_attempts)
```
Comment (casperlamboo) [inline cura/Snapshot.py:208]: I personally would have written this block as the following as I think the `Snapshot._prereadSnapshot(pixel_output)` call is already implicitly incorporated in the `Snapshot.getImageBoundaries(pixel_output)` call
```python
            try:
                min_x, max_x, min_y, max_y = Snapshot.getImageBoundaries(pixel_output)
            except (ValueError, AttributeError) as e:
                if number_of_attempts == 0:
                    Logger.warning( f"Failed to crop the snapshot even after 10 attempts!")
                    return None
                else:
                    number_of_attempts = number_of_attempts - 1
                    Logger.info("Trying to get the snapshot again.")
                    return Snapshot.snapshot(width, height, number_of_attempts)
```

Comment (saumyaj3) [reply cura/Snapshot.py]: oh yes! Then preread function can go (y)

Comment (saumyaj3) [reply cura/Snapshot.py]: [c488164](https://github.com/Ultimaker/Cura/pull/18549/commits/c4881641c3c50add2bfe0b991bc264c82168fb25)

**Pull request merged**